### ISSUE TITLE
Support for SHA256 digest auth (RFC 8760)

### ISSUE THREAD

**Issue by oej** [sipsorcery-org/sipsorcery#525]
I would like to see stronger authentication, as documented in RFC 8760. MD5 is a really poor authentication digest.

**Comment by sipsorcery** [comment]
Hi Olle, thanks for the prompt, I'll take a look at the RFC, it shouldn't be that difficult to implement. 

But the bigger question is how many interop issues is it likely to cause with other SIP stacks? I'm guessing there will be a few out there that don't like multiple digest challenges or that reject a challenge if they don't recognise any of the digests.

For a bit of nostalgia I did attend one of your SIP courses way back in 08/09 in Stockholm.

**Comment by oej** [comment]
Read the RFC and we'll take a discussion over some digital connection. Cool to meet a former student!

**Comment by oej** [comment]
For your info: We're building a new platform with no legacy devices, so I don't really want to implement MD5 auth at all in 2021 when doing that :-)

**Comment by sipsorcery** [comment]
As expected updating the digest logic to use `SHA256` was easy enough. Also as expected, finding interop issues when setting multiple authentication headers was easy. A little bit too easy given that Asterisk is one of the main stream implementations and if it's not yet ready a lot of SIP Providers are not going to be.

This [Asterisk issue](https://issues.asterisk.org/jira/browse/ASTERISK-29397) describes a problem tolerating multiple authentication headers from a UAS. I tested the opposite, sending multiple authentication headers from a UAC. It works if `MD5` is the first header but fails if `SHA-256` is first. I'm capturing the SIP traces for future reference.

**Works if MD5 auth header first**

````
[20:20:59 VRB] INVITE sip:100@192.168.0.48 SIP/2.0
Via: SIP/2.0/UDP 192.168.0.50:52472;branch=z9hG4bK7b2170eb5a144fc197b9f6dd0db231d0;rport
To: <sip:100@192.168.0.48:5060>
From: <sip:7004@192.168.0.48>;tag=UHNRZITEVG
Call-ID: e6820b85cb5c4f30801b138bf09a6800
CSeq: 2 INVITE
Contact: <sip:7004@192.168.0.50:52472>
Max-Forwards: 70
User-Agent: sipsorcery_v5.2.3.0
Authorization: Digest username="7004",realm="asterisk",nonce="1626290287/c3a2522f782e29a3289e93701ba72123",uri="sip:100@192.168.0.48",response="f9d8f41c005d4942047959742d09598c",cnonce="1508316497",nc=00000001,qop=auth,opaque="73db6781466c5493",algorithm=MD5
Authorization: Digest username="7004",realm="asterisk",nonce="1626290287/c3a2522f782e29a3289e93701ba72123",uri="sip:100@192.168.0.48",response="27dc5c60312e2663c31c942f36524f1959cb47d83716c6f95935804f050248a3",cnonce="1239356419",nc=00000001,qop=auth,opaque="73db6781466c5493",algorithm=SHA-256
Supported: replaces, norefersub, 100rel
Content-Length: 260
Content-Type: application/sdp

v=0
o=- 2011314243 0 IN IP4 127.0.0.1
s=sipsorcery
c=IN IP4 192.168.0.50
t=0 0
m=audio 52468 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=sendrecv
a=ssrc:157248748 cname:3f2a15c3-96b1-494b-827e-e7b43bf8155d

[20:20:59 DBG] Response received: udp:0.0.0.0:52472<-udp:192.168.0.48:5060 INVITE 100 Trying
[20:20:59 VRB] SIP/2.0 100 Trying
Via: SIP/2.0/UDP 192.168.0.50:52472;rport=52472;received=192.168.0.50;branch=z9hG4bK7b2170eb5a144fc197b9f6dd0db231d0
To: <sip:100@192.168.0.48>
From: <sip:7004@192.168.0.48>;tag=UHNRZITEVG
Call-ID: e6820b85cb5c4f30801b138bf09a6800
CSeq: 2 INVITE
Content-Length: 0
Server: Asterisk PBX 18.5.0


[20:20:59 DBG] Information response 100 Trying for sip:100@192.168.0.48.
[20:20:59 INF] sip:100@192.168.0.48:5060 Trying: 100 Trying.
[20:20:59 DBG] Response received: udp:0.0.0.0:52472<-udp:192.168.0.48:5060 INVITE 200 OK
[20:20:59 VRB] SIP/2.0 200 OK
Via: SIP/2.0/UDP 192.168.0.50:52472;rport=52472;received=192.168.0.50;branch=z9hG4bK7b2170eb5a144fc197b9f6dd0db231d0
To: <sip:100@192.168.0.48>;tag=df76a369-0374-43a9-82c4-1ed3911262cf
From: <sip:7004@192.168.0.48>;tag=UHNRZITEVG
Call-ID: e6820b85cb5c4f30801b138bf09a6800
CSeq: 2 INVITE
Contact: <sip:192.168.0.48:5060>
Allow: OPTIONS, REGISTER, SUBSCRIBE, NOTIFY, PUBLISH, INVITE, ACK, BYE, CANCEL, UPDATE, PRACK, MESSAGE, REFER
Supported: 100rel, timer, replaces, norefersub
Content-Length: 228
Content-Type: application/sdp
Server: Asterisk PBX 18.5.0

v=0
o=- 2011314243 2 IN IP4 192.168.0.48
s=Asterisk
c=IN IP4 192.168.0.48
t=0 0
m=audio 10620 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv
````

**Fails if SHA256 auth header first**

````
[21:11:49 VRB] INVITE sip:100@192.168.0.48 SIP/2.0
Via: SIP/2.0/UDP 192.168.0.50:62748;branch=z9hG4bK80344b16e6f745c2ba3e022720f89bbc;rport
To: <sip:100@192.168.0.48:5060>
From: <sip:7004@192.168.0.48>;tag=MTTSFRZVWN
Call-ID: d735f80741b24b0f806babaea6db0b3c
CSeq: 2 INVITE
Contact: <sip:7004@192.168.0.50:62748>
Max-Forwards: 70
User-Agent: sipsorcery_v5.2.3.0
Authorization: Digest username="7004",realm="asterisk",nonce="1626293337/3229240d42bdf7ac9e7d22df6fcaf84a",uri="sip:100@192.168.0.48",response="0f5c5751b46da5708d4b7cd3290c5f2997fe7a0d3464526d5b62cae3d8f8079b",cnonce="1443808922",nc=00000001,qop=auth,opaque="774038cb56233792",algorithm=SHA-256
Authorization: Digest username="7004",realm="asterisk",nonce="1626293337/3229240d42bdf7ac9e7d22df6fcaf84a",uri="sip:100@192.168.0.48",response="3a055365dcb9df68deea49e20347c6fb",cnonce="1270560251",nc=00000001,qop=auth,opaque="774038cb56233792",algorithm=MD5
Supported: replaces, norefersub, 100rel
Content-Length: 260
Content-Type: application/sdp

v=0
o=- 1582173763 0 IN IP4 127.0.0.1
s=sipsorcery
c=IN IP4 192.168.0.50
t=0 0
m=audio 62744 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=sendrecv
a=ssrc:696372767 cname:061f3fe8-049b-476f-87ba-d156d8c6177f

[21:11:49 DBG] Response received: udp:0.0.0.0:62748<-udp:192.168.0.48:5060 INVITE 401 Unauthorized
[21:11:49 VRB] SIP/2.0 401 Unauthorized
Via: SIP/2.0/UDP 192.168.0.50:62748;rport=62748;received=192.168.0.50;branch=z9hG4bK80344b16e6f745c2ba3e022720f89bbc
To: <sip:100@192.168.0.48>;tag=z9hG4bK80344b16e6f745c2ba3e022720f89bbc
From: <sip:7004@192.168.0.48>;tag=MTTSFRZVWN
Call-ID: d735f80741b24b0f806babaea6db0b3c
CSeq: 2 INVITE
WWW-Authenticate: Digest realm="asterisk",nonce="1626293337/3229240d42bdf7ac9e7d22df6fcaf84a",qop=auth,opaque="0b9d08f07cbf95ff"
Content-Length: 0
Server: Asterisk PBX 18.5.0
````

**Comment by sipsorcery** [comment]
> For your info: We're building a new platform with no legacy devices, so I don't really want to implement MD5 auth at all in 2021 when doing that :-)

Do you have anything that I can test against? Asterisk (v18.5.0) doesn't seem to have RFC8760 support yet. I'll try FreeSWITCH but a quick Google leads me to believe it might not have it yet either.

It'd be really handy if there was a SIP interop repo on GitHub like the [WebRTC Echoes](https://github.com/sipsorcery/webrtc-echoes) one.

**Comment by oej** [comment]
I have quite often warned about the possible downgrade attacks when doing migration, but the authors of the RFC did not really take that challenge... In my case, since I control both clients and servers, there's only going to be one auth challenge. 

There was some tests at OpenSIPit that gave some feedback on this to the sipcore IETF mailing list, because the developers thought the spec was not clear enough.

**Comment by oej** [comment]
Kamailio doesn't fully support the RFC - in the case of multiple auth headers - but does support SHA256 digest.

**Comment by sipsorcery** [comment]
Do you think this is a reasonable approach?

 - For a `UAC` only send a single authorization header and only use `SHA256` if the `UAS` explicitly set the algorithm in the challenge.
 - For a `UAS` don't set the algorithm but do accept multiple authorization headers with a preference for `SHA256`.

It's not ideal but at least a first step. This is a real chicken-and-egg situation. Until `UAC's` have confidence they can safely send a `SHA256` digest it's too risky, and likewise with `UAS's` needing the confidence to specify `SHA256` in the challenge.

I completely agree `MD5` should be replaced but I can see it causing a lot of pain for users.

**Comment by sipsorcery** [comment]
FreeSWITCH v1.10.6 results.
Tl;dr same as the Asterisk v18.5.0 results.

**Works if MD5 auth header first** (same with single MD5 header)
````
22:52:22 DBG] Response received: udp:0.0.0.0:51549<-udp:192.168.0.47:5060 INVITE 407 Proxy Authentication Required
[22:52:22 VRB] SIP/2.0 407 Proxy Authentication Required
Via: SIP/2.0/UDP 192.168.0.50:51549;branch=z9hG4bK89eaf4c0e0f445e7adfbb17aa8cd26fb;rport=51549
To: <sip:9170@192.168.0.47:5060>;tag=ZyvtttNF4UptH
From: <sip:1000@192.168.0.47>;tag=WVUHNNKFAW
Call-ID: 5a9c1b4bc9214af8a73ae80d70c1e7dd
CSeq: 1 INVITE
User-Agent: FreeSWITCH-mod_sofia/1.10.6-release-18-1ff9d0a60e~64bit
Accept: application/sdp
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, REGISTER, REFER, NOTIFY, PUBLISH, SUBSCRIBE
Proxy-Authenticate: Digest realm="192.168.0.47",nonce="d9920494-56b1-488e-baf3-2f24f9dc1b0e",qop=auth
Supported: timer, path, replaces
Content-Length: 0
Allow-Events: talk, hold, conference, presence, as-feature-event, dialog, line-seize, call-info, sla, include-session-description, presence.winfo, message-summary, refer

[22:52:22 DBG] Response 407 Proxy Authentication Required for sip:9170@192.168.0.47.
[22:52:22 DBG] Request sent: udp:192.168.0.50:51549->udp:192.168.0.47:5060 INVITE sip:9170@192.168.0.47 SIP/2.0
[22:52:22 VRB] INVITE sip:9170@192.168.0.47 SIP/2.0
Via: SIP/2.0/UDP 192.168.0.50:51549;branch=z9hG4bK9a06905b1f424cd5b4bb71d9cf5369f0;rport
To: <sip:9170@192.168.0.47:5060>
From: <sip:1000@192.168.0.47>;tag=WVUHNNKFAW
Call-ID: 5a9c1b4bc9214af8a73ae80d70c1e7dd
CSeq: 2 INVITE
Contact: <sip:1000@192.168.0.50:51549>
Max-Forwards: 70
User-Agent: sipsorcery_v5.2.3.0
Proxy-Authorization: Digest username="1000",realm="192.168.0.47",nonce="d9920494-56b1-488e-baf3-2f24f9dc1b0e",uri="sip:9170@192.168.0.47",response="7feda1b9aad9e230978f8517740ec20e",cnonce="1127420996",nc=00000001,qop=auth,algorithm=MD5
Proxy-Authorization: Digest username="1000",realm="192.168.0.47",nonce="d9920494-56b1-488e-baf3-2f24f9dc1b0e",uri="sip:9170@192.168.0.47",response="27eac217e339c2554eb28917086c371afbdb202811c6b5ff48e6b559d2a6994d",cnonce="1508253422",nc=00000001,qop=auth,algorithm=SHA-256
Supported: replaces, norefersub, 100rel
Content-Length: 261
Content-Type: application/sdp

v=0
o=- 2088992927 0 IN IP4 127.0.0.1
s=sipsorcery
c=IN IP4 192.168.0.50
t=0 0
m=audio 51542 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=sendrecv
a=ssrc:1036439655 cname:eac99574-5990-4afd-adff-e181486d0ac4

[22:52:22 DBG] Information response 100 Trying for sip:9170@192.168.0.47.
[22:52:22 INF] sip:9170@192.168.0.47:5060 Trying: 100 Trying.
[22:52:22 DBG] Response received: udp:0.0.0.0:51549<-udp:192.168.0.47:5060 INVITE 200 OK
[22:52:22 VRB] SIP/2.0 200 OK
Via: SIP/2.0/UDP 192.168.0.50:51549;branch=z9hG4bK9a06905b1f424cd5b4bb71d9cf5369f0;rport=51549
To: <sip:9170@192.168.0.47:5060>;tag=07NKvN6j14cDD
From: <sip:1000@192.168.0.47>;tag=WVUHNNKFAW
Call-ID: 5a9c1b4bc9214af8a73ae80d70c1e7dd
CSeq: 2 INVITE
Contact: <sip:9170@192.168.0.47:5060;transport=udp>
User-Agent: FreeSWITCH-mod_sofia/1.10.6-release-18-1ff9d0a60e~64bit
Accept: application/sdp
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, REGISTER, REFER, NOTIFY, PUBLISH, SUBSCRIBE
Content-Disposition: session
Supported: timer, path, replaces
Content-Length: 220
Content-Type: application/sdp
Allow-Events: talk, hold, conference, presence, as-feature-event, dialog, line-seize, call-info, sla, include-session-description, presence.winfo, message-summary, refer
Remote-Party-ID: "9170" <sip:9170@192.168.0.47>;party=calling;privacy=off;screen=no

v=0
o=FreeSWITCH 1626363474 1626363475 IN IP4 192.168.0.47
s=FreeSWITCH
c=IN IP4 192.168.0.47
t=0 0
m=audio 22468 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
````

**Fails if SHA256 auth header first** (same result if only a single SHA256 auth header is used)
````
[22:48:37 DBG] Response received: udp:0.0.0.0:61912<-udp:192.168.0.47:5060 INVITE 407 Proxy Authentication Required
[22:48:37 VRB] SIP/2.0 407 Proxy Authentication Required
Via: SIP/2.0/UDP 192.168.0.50:61912;branch=z9hG4bKd36a589a4a5f4640929fba0f7aeefc4d;rport=61912
To: <sip:9170@192.168.0.47:5060>;tag=UtQQKej1FrXFK
From: <sip:1000@192.168.0.47>;tag=VZBSXTVYFW
Call-ID: 57977a6d292344cb82ba2d7634de9b36
CSeq: 1 INVITE
User-Agent: FreeSWITCH-mod_sofia/1.10.6-release-18-1ff9d0a60e~64bit
Accept: application/sdp
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, REGISTER, REFER, NOTIFY, PUBLISH, SUBSCRIBE
Proxy-Authenticate: Digest realm="192.168.0.47",nonce="4d03b1eb-62ae-4336-ad79-6669102bfd77",qop=auth
Supported: timer, path, replaces
Content-Length: 0
Allow-Events: talk, hold, conference, presence, as-feature-event, dialog, line-seize, call-info, sla, include-session-description, presence.winfo, message-summary, refer

[22:48:37 DBG] Response 407 Proxy Authentication Required for sip:9170@192.168.0.47.
[22:48:37 DBG] Request sent: udp:192.168.0.50:61912->udp:192.168.0.47:5060 INVITE sip:9170@192.168.0.47 SIP/2.0
[22:48:37 VRB] INVITE sip:9170@192.168.0.47 SIP/2.0
Via: SIP/2.0/UDP 192.168.0.50:61912;branch=z9hG4bKbcd27158d64548608d309b443712e128;rport
To: <sip:9170@192.168.0.47:5060>
From: <sip:1000@192.168.0.47>;tag=VZBSXTVYFW
Call-ID: 57977a6d292344cb82ba2d7634de9b36
CSeq: 2 INVITE
Contact: <sip:1000@192.168.0.50:61912>
Max-Forwards: 70
User-Agent: sipsorcery_v5.2.3.0
Proxy-Authorization: Digest username="1000",realm="192.168.0.47",nonce="4d03b1eb-62ae-4336-ad79-6669102bfd77",uri="sip:9170@192.168.0.47",response="7e5516cee56a8c543b1d886ad248a735e9b01c2e87d5339717b8cc7644d9ff4f",cnonce="1880834677",nc=00000001,qop=auth,algorithm=SHA-256
Proxy-Authorization: Digest username="1000",realm="192.168.0.47",nonce="4d03b1eb-62ae-4336-ad79-6669102bfd77",uri="sip:9170@192.168.0.47",response="8bb5a0ebdfa750f06b638dce7f43c9b9",cnonce="1288191237",nc=00000001,qop=auth,algorithm=MD5
Supported: replaces, norefersub, 100rel
Content-Length: 260
Content-Type: application/sdp

v=0
o=- 1654630854 0 IN IP4 127.0.0.1
s=sipsorcery
c=IN IP4 192.168.0.50
t=0 0
m=audio 61908 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=sendrecv
a=ssrc:516103102 cname:c36e7208-6eee-4202-a16a-84dada7d2b9f

[22:48:37 DBG] Response received: udp:0.0.0.0:61912<-udp:192.168.0.47:5060 INVITE 403 Forbidden
[22:48:37 VRB] SIP/2.0 403 Forbidden
Via: SIP/2.0/UDP 192.168.0.50:61912;branch=z9hG4bKbcd27158d64548608d309b443712e128;rport=61912
To: <sip:9170@192.168.0.47:5060>;tag=v3ggN924c1K2e
From: <sip:1000@192.168.0.47>;tag=VZBSXTVYFW
Call-ID: 57977a6d292344cb82ba2d7634de9b36
CSeq: 2 INVITE
User-Agent: FreeSWITCH-mod_sofia/1.10.6-release-18-1ff9d0a60e~64bit
Accept: application/sdp
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, REGISTER, REFER, NOTIFY, PUBLISH, SUBSCRIBE
Supported: timer, path, replaces
Content-Length: 0
Allow-Events: talk, hold, conference, presence, as-feature-event, dialog, line-seize, call-info, sla, include-session-description, presence.winfo, message-summary, refer
````
